course4/homework-unique-pointer/collectable.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+#include<iostream>
+#include<vector>
+
+namespace course{
+
+template <typename T>
+class CollectablePointer
+{
+private:
+    T* payload;
+    static std::vector<T*> poolOfDead;
+    CollectablePointer() = default;
+
+public:
+    CollectablePointer(const CollectablePointer& other)
+    {
+        payload = new T(*other.payload);
+        std::cout << "allocated as copy: " << (void*)payload << std::endl;
+    }
+
+    CollectablePointer(CollectablePointer&& other)
+    {
+        payload = std::move(other.payload);
+        other.payload = nullptr;
+    }
+
+    ~CollectablePointer()
+    {
+        if(payload != nullptr)
+        {
+            CollectablePointer<T>::poolOfDead.push_back(payload);
+            payload = nullptr;
+        }
+    }
+
+    T& get()
```

Comment:
Well design smart pointer should have intereface as similar to a raw pointer as possible.
Therefore it should have at least overloaded operator ```*``` and ```->```.